theatre/studio/src/utils/derive-utils.tsx

```diff
@@ -0,0 +1,117 @@
+import {isDerivation, prism} from '@theatre/dataverse'
+import type {IDerivation} from '@theatre/dataverse'
+import {useDerivation} from '@theatre/react'
+import type {$IntentionalAny} from '@theatre/shared/utils/types'
+import React, {useMemo, useRef} from 'react'
+import {invariant} from './invariant'
+
+type DeriveAll<T> = IDerivation<
+  {
+    [P in keyof T]: T[P] extends IDerivation<infer R> ? R : never
+  }
+>
+export function deriveAllD<T extends [...IDerivation<any>[]]>(
+  obj: T,
+): DeriveAll<T>
+export function deriveAllD<T extends Record<string, IDerivation<any>>>(
+  obj: T,
+): DeriveAll<T>
+export function deriveAllD(
+  obj: Record<string, IDerivation<any>> | IDerivation<any>[],
+): DeriveAll<Record<string, IDerivation<any>> | IDerivation<any>[]> {
+  return prism(() => {
+    if (Array.isArray(obj)) {
+      const values = new Array(obj.length)
+      for (let i = 0; i < obj.length; i++) {
+        values[i] = obj[i].getValue()
+      }
+      return values
+    } else {
+      const values: $IntentionalAny = {}
+      for (const k in obj) {
+        values[k] = obj[k].getValue()
+      }
+      return values
+    }
+  }) as $IntentionalAny
+}
+
+export function deriveReact<T>(
+  der: IDerivation<T>,
+  render: DeriveRenderFn<T>,
+): JSX.Element {
+  return <DeriveElement der={der} render={render} />
+}
+
+export function deriveChild(der: IDerivation<ValidJSXChild>): JSX.Element {
+  return <DeriveElement der={der} render={undefined} />
+}
+
+export type ValidJSXChild = React.ReactNode | JSX.Element | JSX.Element[]
+export type DeriveRenderFn<T> =
+  | (T extends ValidJSXChild ? void : never)
+  | ((value: T) => ValidJSXChild)
+
+function DeriveElement<T>(props: {
+  der: IDerivation<T>
+  render: DeriveRenderFn<T>
+}) {
+  const value = useDerivation(props.der)
+  return <>{props.render != null ? props.render(value) : value}</>
+}
+
+class TSErrors<M> extends Error {}
+
+type ReactD<Props extends {}> = (
+  props: {
+    [P in keyof Props]: Props[P] extends IDerivation<infer _>
+      ? TSErrors<"Can't both use Derivation properties while wrapping with reactDerived">
+      : Props[P] | IDerivation<Props[P]>
+  },
+) => React.ReactElement | null
+
+/**
+ * Wrap up the component to enable it to take derivable properties.
+ * Invoked similarly to `React.memo`.
+ */
+export function reactDerived<Props extends {}>(
```

Comment:
- Naming nitpick: technically, the component is not the one that is derived, it just observes its props. We don't have a term for observers in the dataverse lingo yet though, but changing it to reflect that would be nicer.

- I'm not a fan of namespacing in export names. As long as the name is unambiguous in the module context, it's fine, we can always alias it at import. `reactDerived` is also easily confused with `deriveReact`.